rllib/evaluation/rollout_worker.py

```diff
@@ -1392,8 +1392,8 @@ def foreach_policy_to_train(
             `func([policy, pid, **kwargs])`.
         """
         return [
-            func(policy, pid, **kwargs)
-            for pid, policy in self.policy_map.items()
+            func(self.policy_map[pid], pid, **kwargs)
+            for pid in self.policy_map.keys()
```

Comment:
I can't really see the difference between current and updated version here.
can you say a few words about the intention here?
same below in get_weights() function.

Reply:
Yes from the view of the `rollout_worker` the behavior shouldn't change.
The difference is that iterating over `items()` will access both the key and the policy, i.e. internally `PolicyMap` will call `__getitem__`. `__getitem__` will access the disk, if the selected policy is not in the cache. Every policy is accessed and then *afterwards* the if statement in the next line checks whether we are actually interested in this policy. With my changes we only call `__getitem__` for those policies that pass through the if statement. This way we reduce the number of unnecessary access of policies and in turn reduce the number of unnecessary `_read_from_disk()` (Compare `rllib/policy/policy_map.py`).
The same goes for `get_weights()`.

In short: Don't do `policy_map[pid]` if we skip this `pid` anyway, due to the if statement.

Reply:
🤯 this is some highly optimized stuff.
can you please help add a big comment above these 2 places, about what you just said, so we don't come in and accidentally "clean up" these loops some day.
appreciate the thoughtful changes.

Reply:
understand! this is great.
I think global_vars are mostly for training purpose, so it should be fine to skip if a policy is not being trained.